asserts/snapasserts/validation_sets.go

```diff
@@ -0,0 +1,244 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapasserts
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/snap"
+)
+
+// ValidationSetsConflictError describes an error where multiple
+// validation sets are in conflict about snaps.
+type ValidationSetsConflictError struct {
+	Sets  map[string]*asserts.ValidationSet
+	Snaps map[string]error
+}
+
+func (e *ValidationSetsConflictError) Error() string {
+	buf := bytes.NewBufferString("validation sets are in conflict:")
+	for _, err := range e.Snaps {
+		fmt.Fprintf(buf, "\n- %v", err)
+	}
+	return buf.String()
+}
+
+// ValidationSets can hold a combination of validation-set assertions
+// and can check for conflicts or help applying them.
+type ValidationSets struct {
+	// sets maps sequence keys to validation-set in the combination
+	sets map[string]*asserts.ValidationSet
+	// snaps maps snap-ids to snap constraints
+	snaps map[string]*snapContraints
+}
+
+const presConflict asserts.Presence = "conflict"
+
+var unspecifiedRevision = snap.R(0)
+var invalidPresRevision = snap.R(-1)
+
+type snapContraints struct {
+	name     string
+	presence asserts.Presence
+	// revisions maps revisions to pairing of ValidationSetSnap
+	// and the origination validation-set key
+	// * unspecifiedRevision is used for constraints without a
+	//   revision
+	// * invalidPresRevision is used for constraints that mark
+	//   presence as invalid
+	revisions map[snap.Revision][]*revConstraint
+}
+
+type revConstraint struct {
+	validationSetKey string
+	asserts.ValidationSetSnap
+}
+
+func (c *snapContraints) conflict() (conflicting []string, err error) {
```

Comment:
This function has unusual signature. I suspect it would be less surprising to return this instead:
```suggestion
func (c *snapContraints) conflict() (conflicting []string, errMsg string) {
```

This way the caller may well expect to return both `conflicting` and `errMsg`.

Reply:
I did something slightly different in the end, I made conflict() return a snapConflictsError carrying the sets found by whichSets here and deferred some of the message formatting to snapConflictsError.Error